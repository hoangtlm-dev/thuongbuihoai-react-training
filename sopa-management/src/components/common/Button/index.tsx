import React, { memo } from "react";
import "./button.css";

export enum ButtonType {
  default = "",
  primary = "primary",
  secondary = "secondary",
  tertiary = "tertiary",
  btnIconPrimary = "icon-primary",
  btnIconSecondary = "icon-secondary"
};

interface ButtonProps {
  children?: string | JSX.Element;
  type?: ButtonType;
  disable?: boolean;
  onClick?: () => void;
  onSubmit?: () => void;
};

const Button: React.FC<ButtonProps> = ({
  children = "",
  type = ButtonType.default,
  disable = false,
  onClick = () => {},
  onSubmit = () => {},
}) => {
  let className = "btn";
  switch (type) {
    case ButtonType.primary:
      className += ` btn-${ButtonType.primary}`;
      break;
    case ButtonType.secondary:
      className += ` btn-${ButtonType.secondary}`;
      break;
    case ButtonType.tertiary:
      className += ` btn-${ButtonType.tertiary}`;
      break;
    case ButtonType.btnIconPrimary:
      className += ` btn-${ButtonType.btnIconPrimary}`;
      break;
    case ButtonType.btnIconSecondary:
      className += ` btn-${ButtonType.btnIconSecondary}`;
      break;
    default:
      break;
  };

  return (
    <button
      data-testid="button"
      className={`${className} btn-${type}`}
      disabled={disable}
      onClick={onClick}
      onSubmit={onSubmit}
    >{children}</button>
  );
};

export default memo(Button);
