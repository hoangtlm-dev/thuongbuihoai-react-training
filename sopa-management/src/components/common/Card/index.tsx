import { memo } from 'react';
import { Link } from 'react-router-dom';

// Components
import Text,
{
  ThemeType
} from '@common/Text';
import Price,
{
  PriceType
} from '@common/Price';
import Image from '@common/Image';

// Styles
import './card.css';

interface CardProps {
  key?: string;
  name?: string;
  color?: string;
  price?: number;
  href: string;
  src: string;
  alt?: string;
}

const Card: React.FC<CardProps> = ({
  name = '',
  color = '',
  price = 0,
  href,
  src,
  alt = ''
}) => {
  return (
    <div data-testid='card' className='card'>
      <div className='card-item'>
        <figure className='card-image'>
          <Image href={href} url={src} alt={alt} className='card-image-item' />
        </figure>
        <div className='card-info'>
          <Link to={href} className='card-name'>
            {name}
          </Link>
          <div className='card-description'>
            <Text text={color} type={ThemeType.highlightPrimary} />
            <Price value={price} type={PriceType.primary} />
          </div>
        </div>
      </div>
    </div>
  );
};

export default memo(Card);
