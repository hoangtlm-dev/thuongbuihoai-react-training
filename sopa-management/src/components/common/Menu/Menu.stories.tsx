import type { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Menu, { MenuTheme } from '.';

// Constants
import {
  MENU_EVERYTHING_ELSE,
  MENU_HEADER
} from '../../../constants/common';

const meta: Meta<typeof Menu> = {
  component: Menu,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
};

export default meta;
type Story = StoryObj<typeof Menu>;

export const MenuHeader: Story = {
  args: {
    menuList: MENU_HEADER,
    type: MenuTheme.primary
  }
};

export const MenuFooter: Story = {
  args: {
    mainItem: "Everything Else",
    menuList: MENU_EVERYTHING_ELSE,
    type: MenuTheme.secondary
  }
};

