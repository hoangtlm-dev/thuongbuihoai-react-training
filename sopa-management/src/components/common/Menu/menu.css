.menu-list-vertical {
  display: flex;
}

.menu-item-vertical {
  list-style: none;
  font-size: var(--text-sm);
  margin-right: 24px;
}

.menu-item-vertical > a {
  font-size: var(--text-extra-sm);
  text-decoration: none;
  color: var(--dark);
}

.menu-main-horizontal {
  margin-bottom: 20px;
}

.menu-list-horizontal {
  display: flex;
  flex-direction: column;
}

.menu-item-horizontal {
  list-style: none;
  font-size: var(--text-sm);
  margin-bottom: 20px;
}

.menu-item-horizontal > a,
.menu-main-horizontal {
  font-size: var(--text-extra-sm);
  text-decoration: none;
  color: var(--light);
}

.menu-main-horizontal {
  font-size: var(--text-base);
  font-weight: var(--font-medium);
}
