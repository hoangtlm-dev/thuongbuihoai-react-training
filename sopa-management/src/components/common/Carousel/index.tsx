// Images
import { IMAGE } from '@constants/image';

// Components
import Button,
{
  ButtonType
} from '@common/Button';
import Image from '@common/Image';
import Input from '@common/Input';
import Text,
{
  SizeType
} from '@common/Text';

// Constants
import { CAROUSEL_BULLET } from '@constants/common';

// Styles
import './carousel.css';

const Carousel: React.FC = () => (
  <div data-testid='carousel' className='carousel'>
    <div className='carousel-inner'>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-1'
        name='carousel'
        hidden
        checked
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner} alt={IMAGE.altBanner} />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-2'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner2} alt={IMAGE.altBanner} />
      </figure>
      <Input
        className='carousel-open'
        type='radio'
        id='carousel-3'
        name='carousel'
        hidden
      />
      <figure className='carousel-item'>
        <Image url={IMAGE.banner3} alt={IMAGE.altBanner} />
      </figure>
      <ul className='carousel-indicators'>
        {CAROUSEL_BULLET.map((item) => (
          <li key={item.key}>
            <label htmlFor={item.htmlFor} className={item.className} />
          </li>
        ))}
      </ul>
    </div>
    <div className='carousel-info'>
      <Text
        text='Step inside, for comfort and magic await you.'
        type={SizeType.large}
        className='carousel-heading'
      />
      <Button children='Show now' type={ButtonType.primary} />
    </div>
  </div>
);

export default Carousel;
