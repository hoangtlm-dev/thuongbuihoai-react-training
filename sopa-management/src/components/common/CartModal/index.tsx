import { memo } from 'react';

// Images
import { Close } from '../../../../public/images/icons';

// Components
import Text from '../Text';
import Button,
{
  ButtonType
} from '../Button';
import Price,
{
  PriceType
} from '../Price';

// Interfaces
import { Product } from '../../../interfaces/product';

// Helpers
import { totalPrices } from '../../../helpers/common';

// Stores
import { useCartStore } from '../../../stores/cart';

// Styles
import './cartModal.css';
import CartItem from './CartItem';

interface CartModalProps {
  onToggleModal?: () => void;
}
const CartModal: React.FC<CartModalProps> = ({
  onToggleModal = () => {},
}) => {
  const { carts } = useCartStore();

  return (
    <div data-testid='cart-modal' className='overlay'>
      <div className='cart'>
        <div className='cart-header'>
          <Text text='Cart' className='cart-text' />
          <Button
            children={<Close />}
            type={ButtonType.btnIconPrimary}
            onClick={onToggleModal}
          />
        </div>
        <div className='card-body'>
          {carts?.map((cartItem: Product) => (
            <CartItem
              key={cartItem.id}
              cartItem={cartItem}
            />
          ))}
        </div>
        <hr />
        <div className='cart-footer'>
          <Text text='Subtotal' className='cart-text' />
          <Price value={totalPrices(carts)} type={PriceType.tertiary} />
        </div>
      </div>
    </div>
  );
};

export default memo(CartModal);
