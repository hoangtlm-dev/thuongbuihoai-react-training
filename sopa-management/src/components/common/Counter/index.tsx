import { memo } from 'react';

// Components
import Button, { ButtonType } from '@common/Button';
import Input from '../Input';

// Styles
import './counter.css';

interface CounterProps {
  value: number;
  onIncrement: () => void;
  onDecrement: () => void;
}

const Counter = ({
  value,
  onIncrement,
  onDecrement
}: CounterProps) => (
  <div data-testid='counter' className='counter'>
    <Button
      ariaLabel='Decrement'
      type={ButtonType.btnIconPrimary}
      className='counter-minus'
      onClick={onDecrement}
    />
    <Input type='number' classNameInput='counter-input' value={value} />
    <Button
      ariaLabel='Increment'
      type={ButtonType.btnIconPrimary}
      className='counter-plus'
      onClick={onIncrement}
    />
  </div>
);

export default memo(Counter);
