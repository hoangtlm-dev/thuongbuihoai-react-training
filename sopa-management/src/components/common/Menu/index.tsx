import React, { memo } from 'react';
import { Link } from 'react-router-dom';

// Styles
import './menu.css';

export type MenuType = {
  key: string;
  label: string;
};

export enum MenuTheme {
  default = '',
  vertical = 'vertical',
  horizontal = 'horizontal'
}

export interface MenuProps {
  menuList: MenuType[];
  menuFooter?: string;
  type?: string;
  className?: string;
}

const Menu: React.FC<MenuProps> = memo(
  ({
    menuList,
    menuFooter,
    type,
    className
  }) => {
    const renderMenuHeader = (list: MenuType[]) => {
      return list.map((item) => (
        <li className={`${className} menu-item-${type}`} key={item.key}>
          <Link to='#'>{item.label}</Link>
        </li>
      ));
    };

    return (
      <div data-testid='menu' className='menu'>
        <ul className={`${className} menu-list-${type}`}>
          {menuFooter && (
            <p className={`menu-main-${type}`}>{menuFooter}</p>
          )}
          {renderMenuHeader(menuList)}
        </ul>
      </div>
    );
  }
);

export default Menu;
