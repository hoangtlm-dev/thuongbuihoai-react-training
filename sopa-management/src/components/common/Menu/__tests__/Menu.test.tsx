// Libs
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';

// Components
import Menu from '../index';

// Constants
import { MENU_HEADER } from '../../../../constants/common';

describe('Menu component', () => {
  test('Should render Menu component by default', () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <Menu menuList={MENU_HEADER} />
      </Router>
    );
    const menu = getByTestId('menu');
    expect(menu).toBeInTheDocument();
  });

  test('Should render Menu component by value', () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <Menu menuList={MENU_HEADER} value='Products' />
      </Router>
    );
    const menu = getByTestId('menu');
    expect(menu).toBeInTheDocument();
  });

  test('Matches snapshot', () => {
    const history = createMemoryHistory();
    const { container } = render(
      <Router location={history.location} navigator={history}>
        <Menu menuList={MENU_HEADER} />
      </Router>
    );
    expect(container).toMatchSnapshot();
  });
});
