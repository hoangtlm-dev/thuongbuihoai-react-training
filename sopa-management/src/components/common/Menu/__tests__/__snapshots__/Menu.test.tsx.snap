// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component Matches snapshot 1`] = `
<div>
  <div
    class="menu"
    data-testid="menu"
  >
    <ul
      class="undefined menu-list-undefined"
    >
      <li
        class="undefined menu-item-undefined"
      >
        <a
          href="/"
        >
          Men
        </a>
      </li>
      <li
        class="undefined menu-item-undefined"
      >
        <a
          href="/"
        >
          Women
        </a>
      </li>
      <li
        class="undefined menu-item-undefined"
      >
        <a
          href="/"
        >
          Accessories
        </a>
      </li>
      <li
        class="undefined menu-item-undefined"
      >
        <a
          href="/"
        >
          About
        </a>
      </li>
    </ul>
  </div>
</div>
`;
